crates/bevy_core_pipeline/src/taa/taa.wgsl

```diff
@@ -0,0 +1,190 @@
+    // https://advances.realtimerendering.com/s2014/index.html#_HIGH-QUALITY_TEMPORAL_SUPERSAMPLING, slide 27
+    let offset = texel_size * 2.0;
+    let d_uv_tl = uv + vec2(-offset.x, offset.y);
+    let d_uv_tr = uv + vec2(offset.x, offset.y);
+    let d_uv_bl = uv + vec2(-offset.x, -offset.y);
+    let d_uv_br = uv + vec2(offset.x, -offset.y);
+    var closest_uv = uv;
+    let d_tl = textureSample(depth, nearest_sampler, d_uv_tl);
+    let d_tr = textureSample(depth, nearest_sampler, d_uv_tr);
+    var closest_depth = textureSample(depth, nearest_sampler, uv);
+    let d_bl = textureSample(depth, nearest_sampler, d_uv_bl);
+    let d_br = textureSample(depth, nearest_sampler, d_uv_br);
+    if d_tl > closest_depth {
+        closest_uv = d_uv_tl;
+        closest_depth = d_tl;
+    }
+    if d_tr > closest_depth {
+        closest_uv = d_uv_tr;
+        closest_depth = d_tr;
+    }
+    if d_bl > closest_depth {
+        closest_uv = d_uv_bl;
+        closest_depth = d_bl;
+    }
+    if d_br > closest_depth {
+        closest_uv = d_uv_br;
+    }
+    let closest_motion_vector = textureSample(motion_vectors, nearest_sampler, closest_uv).rg;
+
+    // Reproject to find the equivalent sample from the past
+    // Uses 5-sample Catmull-Rom filtering (reduces blurriness)
+    // https://gist.github.com/TheRealMJP/c83b8c0f46b63f3a88a5986f4fa982b1
+    // https://vec3.ca/bicubic-filtering-in-fewer-taps
+    // https://developer.nvidia.com/gpugems/gpugems2/part-iii-high-quality-rendering/chapter-20-fast-third-order-texture-filtering
+    // https://www.activision.com/cdn/research/Dynamic_Temporal_Antialiasing_and_Upsampling_in_Call_of_Duty_v4.pdf#page=68
+    let history_uv = uv - closest_motion_vector;
+    let sample_position = history_uv * texture_size;
+    let texel_center = floor(sample_position - 0.5) + 0.5;
+    let f = sample_position - texel_center;
+    let w0 = f * (-0.5 + f * (1.0 - 0.5 * f));
+    let w1 = 1.0 + f * f * (-2.5 + 1.5 * f);
+    let w2 = f * (0.5 + f * (2.0 - 1.5 * f));
+    let w3 = f * f * (-0.5 + 0.5 * f);
+    let w12 = w1 + w2;
+    let texel_position_0 = (texel_center - 1.0) * texel_size;
+    let texel_position_3 = (texel_center + 2.0) * texel_size;
+    let texel_position_12 = (texel_center + (w2 / w12)) * texel_size;
+    var history_color = sample_history(texel_position_12.x, texel_position_0.y) * w12.x * w0.y;
+    history_color += sample_history(texel_position_0.x, texel_position_12.y) * w0.x * w12.y;
+    history_color += sample_history(texel_position_12.x, texel_position_12.y) * w12.x * w12.y;
+    history_color += sample_history(texel_position_3.x, texel_position_12.y) * w3.x * w12.y;
+    history_color += sample_history(texel_position_12.x, texel_position_3.y) * w12.x * w3.y;
+
+    // Constrain past sample with 3x3 YCoCg variance clipping (reduces ghosting)
+    // YCoCg: https://advances.realtimerendering.com/s2014/index.html#_HIGH-QUALITY_TEMPORAL_SUPERSAMPLING, slide 33
```

Comment:
Slide 32 suggests using a cross-shape to 'round off the filter' because the 3x3 box introduces artifacts. Did you try that?

Reply:
Unreal uses basic clipping, and not variance clipping like my implementation. I don't think slide 32 applies.